components/engine_rocks/src/util.rs

```diff
@@ -331,6 +333,137 @@ pub fn from_raw_perf_level(level: rocksdb::PerfLevel) -> engine_traits::PerfLeve
     }
 }
 
+struct KeyRange {
+    start_key: Box<[u8]>,
+    end_key: Box<[u8]>,
+}
+
+pub struct NoopFilter;
+
+impl CompactionFilter for NoopFilter {}
+
+pub struct NoopFactory;
+
+impl CompactionFilterFactory for NoopFactory {
+    type Filter = NoopFilter;
+    fn create_compaction_filter(
+        &self,
+        _context: &CompactionFilterContext,
+    ) -> Option<(CString, Self::Filter)> {
+        None
+    }
+
+    fn should_filter_table_file_creation(&self, reason: DBTableFileCreationReason) -> bool {
+        matches!(
+            reason,
+            DBTableFileCreationReason::Flush | DBTableFileCreationReason::Compaction
+        )
+    }
+}
+
+/// Used to build `RangeCompactionFilterFactory`, optionally stacked over
+/// another compaction filter factory.
+pub struct RangeCompactionFilterFactoryBuilder(Arc<KeyRange>);
+
+impl RangeCompactionFilterFactoryBuilder {
+    pub fn new(start_key: Vec<u8>, end_key: Vec<u8>) -> Self {
+        Self(Arc::new(KeyRange {
+            start_key: start_key.into_boxed_slice(),
+            end_key: end_key.into_boxed_slice(),
+        }))
+    }
+
+    pub fn build_with<C: CompactionFilterFactory>(
+        &self,
+        inner: C,
+    ) -> RangeCompactionFilterFactory<C> {
+        assert!(inner.should_filter_table_file_creation(DBTableFileCreationReason::Compaction));
```

Comment:
The assert seems useless. It does no harm to filter all those reasons.

Reply:
The outer filter shadows the `should_filter` implementation of inner filter. It's needed to make sure we don't mistakenly feed the inner filter something it doesn't expect.

About the "filter out" thing, I'll have to fix all the broken tests that assume flush isn't filtered before doing that.

Reply:
But this is not extensible. For example, if we want to add a new compaction filter to wrap `RangeCompactionFilterFactory`, then we need another testing and asserts. We can grep `CompactionFilterFactory for` to find out all implementation and filter out unwanted reason in `create_compaction_filter`.

Reply:
I decoupled the stacking part into a separate filter.